cpp/biconnected_components_module/algorithm/biconnected_components.cpp

```diff
@@ -0,0 +1,75 @@
+#include "biconnected_components.hpp"
+
+void bcc_utility::BccDFS(std::uint64_t node_id, std::uint64_t parent_id, bcc_utility::NodeState *state,
+                         std::stack<mg_graph::Edge<>> *edge_stack, std::vector<std::vector<mg_graph::Edge<>>> *bcc,
+                         const mg_graph::GraphView<> *graph) {
+  auto root = node_id == parent_id;
+
+  state->counter++;
+  state->visited[node_id] = true;
+  state->discovery[node_id] = state->counter;
+  state->low_link[node_id] = state->counter;
+
+  std::uint64_t root_count = 0;  // needed to handle the special case for root node.
+  for (const auto &neigh : graph->Neighbours(node_id)) {
+    auto next_id = neigh.node_id;
+
+    const auto &edge = graph->GetEdge(neigh.edge_id);
+    if (state->visited[next_id]) {
+      if (next_id != parent_id) {
+        state->low_link[node_id] = std::min(state->low_link[node_id], state->discovery[next_id]);
+        if (state->discovery[next_id] < state->discovery[node_id]) edge_stack->push(edge);
+      }
+      continue;
+    }
+
+    ++root_count;
+    edge_stack->push(edge);
+    BccDFS(next_id, node_id, state, edge_stack, bcc, graph);
+    state->low_link[node_id] = std::min(state->low_link[node_id], state->low_link[next_id]);
+
+    // Articulation point check
+    if ((root && root_count > 1) ||  // special case for root
+        (!root && state->low_link[next_id] >= state->discovery[node_id])) {
+      // obsdata::Edges until (node, next) on stack form a BCC
+      bcc->emplace_back();
+      while (edge_stack->top().id != edge.id) {
```

Comment:
What if the stack is empty? Can it happen?

Reply:
Since Dino was the author of this algorithm I am not sure, because I haven't jumped into implementation details. However, in line 27, there is adding into a stack which means it will not be empty when recursion continues after line 29. Finally, I will include a check of not being empty in the if-clause.

Reply:
Oh, I see, we have a guarantee that at least the lowest member will be `== edge.id`.